### PR TITLE
Animated webp support

### PR DESCRIPTION
**Pull request by december1981**
As well as supporting animated webp, this change allows you to download a resource from a url, eg https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExcXVtOWhqNTc0dmFxY3BuaWhwYzh2ZnM5cWwxNWdmdWtyMG1zejBnZiZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/Vbu1MROenErCauNAet/giphy.gif

### CODE REVIEW

File: src/pipeline/frames.rs [hunk 1]
```diff
@@ -310,7 +364,65 @@ fn open_gif(path: &Path) -> Result<FrameIterator, MyError> {
...
+    let mut frames = Vec::new();
+    // this code is based on the code example here:
+    // https://developers.google.com/speed/webp/docs/container-api#webpanimdecoder_api
+    unsafe {
+        let mut options = webp::WebPAnimDecoderOptions{
+            color_mode: webp::WEBP_CSP_MODE::MODE_RGBA,
+            use_threads: 0,
+            padding: [0, 0, 0, 0, 0, 0, 0],
+        };
+        webp::WebPAnimDecoderOptionsInit(&mut options);
```
Comment (december1981) [inline src/pipeline/frames.rs:400]: could experiment with use_threads set to 1 to make decoding faster.

File: src/pipeline/frames.rs [hunk 2]
```diff
@@ -66,15 +67,16 @@ impl Iterator for FrameIterator {
         match self {
             FrameIterator::Image(ref mut img) => img.take(),
             FrameIterator::Video(ref mut video) => capture_video_frame(video),
-            FrameIterator::AnimatedGif {
+            FrameIterator::AnimatedImage {
                 ref frames,
                 ref mut current_frame,
             } => {
-                if *current_frame == frames.len() - 1 {
+                if *current_frame == frames.len() {
```
Comment (december1981) [inline src/pipeline/frames.rs:74]: caught this regression issue (my bad, sorry) when I tried a single frame webp

**Comment by december1981** [comment]
As an aside, you'll note the gif example I gave has image artifacts on replay. If you download it, and convert it to webp with ffmpeg

`ffmpeg -f gif -i wakeup.gif -c libwebp .wakeup.webp`

And run the webp version, you'll see the image is clean. This maybe an issue with the gif crate.

The issue is more egregious with this gif:
https://media.giphy.com/media/v1.Y2lkPTc5MGI3NjExa3M2eXdpd3F4czVleWx6Y28zcW5kMGg1M3Z2YWpiMWZrdGFtcmFicyZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/TLL98D1dhycYOnEtT7/giphy.gif

**Comment by december1981** [comment]
I guess one thing to improve is using the timestamp information for a AnmatedImage in the frames list, to control the replay speed. The wakeup webp I had to play at --fps 12 to see it at the expected rate.

**Comment by maxcurzi** [comment]
thanks for this MR, I'll check it out after the Easter holidays :)

**Comment by december1981** [comment]
I've resolved the issues with gif rendering and fps. The gif decoder phase required another layer for the frame data to be composed correctly. However, I note that you can just pass the gif file directly to `open_video`, and opencv treats it perfectly. (You might consider removing the gif crate entirely, although I think there's merit in hand decoding this format.)

File: src/pipeline/frames.rs (hunk 2 above)
Comment (maxcurzi) [reply src/pipeline/frames.rs:74]: not sure it was this line but well done, the previous version paused for a frame at the end of a gif . Reverting this change now skips ahead one frame which I think it's due to the subtracting one frame time when looping. Seems fine now

**Comment by maxcurzi** [comment]
interesting how the blockiness introduced by the ffmpeg conversion to webp reduced the noise in the ascii version
left gif (1.9MB), right webp (570KB)
image

I tried to treat gifs and webp as videos in open_media_from_path but the resulting terminal output was 
`[webp @ 0x5646b4ad1580] image data not found`

Not 100% sure why, though. 

The main thing is that while it may work just fine I wanted sooner or later to move away from OpenCV because it makes installing this crate non-trivial. Ideally you'd want to do `cargo install tplay` and everything works, but the OpenCV dependency makes this crate harder to port / install / use on Windows/Mac.

I haven't had much time to investigate simpler alternatives, though I can believe that OpenCV would be hard to beat in terms of flexibility and performance.

Changes look good!

**Comment by december1981** [comment]
Thanks for merging. Yes, I see what you mean about OpenCV. If I've understood its capabilities, it also carries all sorts of extraneous features not required for this application, like image recognition.

**Pull request merged**